alpa/serialization.py

```diff
@@ -28,169 +26,134 @@
-            save_dir = _get_save_path()
-            ts_store(save_dir, x)
-            return msgpack.ExtType(_MsgpackExtType.ndarray,
-                                   msgpack.packb(save_dir))
-        if np.issctype(type(x)):
-            # pack scalar as ndarray
-            return msgpack.ExtType(_MsgpackExtType.npscalar,
-                                   _ndarray_to_bytes(np.asarray(x)))
-        elif isinstance(x, complex):
-            return msgpack.ExtType(_MsgpackExtType.native_complex,
-                                   msgpack.packb((x.real, x.imag)))
-        elif isinstance(x, DistributedArray):
-            save_dir = _get_save_path()
-            x.save(save_dir)
-            return msgpack.ExtType(_MsgpackExtType.distarray,
-                                   msgpack.packb(save_dir))
-        elif isinstance(x, ReplicatedDistributedArray):
-            save_dir = _get_save_path()
-            x.replica.save(save_dir)
-            return msgpack.ExtType(_MsgpackExtType.replicated_distarray,
-                                   msgpack.packb(save_dir))
-        return x
-
-    return _msgpack_ext_pack
-
-
-def _msgpack_ext_unpack(code, data):
-    """Messagepack decoders for custom types."""
-    if code == _MsgpackExtType.ndarray:
-        return msgpack.unpackb(data)
-    elif code == _MsgpackExtType.native_complex:
-        complex_tuple = msgpack.unpackb(data)
-        return complex(complex_tuple[0], complex_tuple[1])
-    elif code == _MsgpackExtType.npscalar:
-        ar = _ndarray_from_bytes(data)
-        return ar[()]  # unpack ndarray to scalar
-    elif code == _MsgpackExtType.distarray:
-        return msgpack.unpackb(data)
-    elif code == _MsgpackExtType.replicated_distarray:
-        return msgpack.unpackb(data)
-    return msgpack.ExtType(code, data)
-
-
-def get_ts_spec(ckpt_path: str):
-    spec = {
-        'driver': 'zarr',
-        'kvstore': {},
-        'metadata_key': ".zarray0"
-    }
-    if ckpt_path.startswith('gs://'):
-        m = re.fullmatch('^gs://([^/]*)/(.*)$', ckpt_path, re.DOTALL)
-        if m is None:
-            raise ValueError(
-                'The ckpt_path should contain the bucket name and the '
-                f'file path inside the bucket. Got: {ckpt_path}')
-        gcs_bucket = m.group(1)
-        path_without_bucket = m.group(2)
-        spec['kvstore'] = {
-            'driver': 'gcs',
-            'bucket': gcs_bucket,
-            'path': path_without_bucket
-        }
+def _dfs_pytree(tree, prefix, paths):
+    if isinstance(tree, dict):
+        for k, v in tree.items():
+            _dfs_pytree(v, prefix + "." + str(k), paths)
+    elif isinstance(tree, (tuple, list)):
+        for i, v in enumerate(tree):
+            _dfs_pytree(v, prefix + "." + str(i), paths)
+    elif tree is None:
+        return
     else:
-        spec['kvstore'] = {'driver': 'file', 'path': ckpt_path}
-    return spec
+        # Leaf node
+        paths.append(prefix)
```

Comment:
Cite from [Jax's Doc](https://jax.readthedocs.io/en/latest/pytrees.html):
By default, pytree containers can be lists, tuples, dicts, namedtuple, **None**, OrderedDict. Other types of values, including numeric and ndarray values, are treated as leaves.
Since there is not way to further flatten the None container, I simply skip it.